### PR TITLE
Add Policy Check Handler to check if packages contain prettier 

### PR DESCRIPTION
**Pull request by jikim-msft**
This PR adds a policy-check handler which checks that prettier exists in every `package.json`

### CODE REVIEW

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts [hunk 1]
```diff
@@ -585,6 +585,30 @@ export const handlers: Handler[] = [
...
+            const hasPrettierProperty = Object.prototype.hasOwnProperty.call(json, "prettier");
+            const hasPrettierFixProperty = Object.prototype.hasOwnProperty.call(
+                json,
+                "prettier:fix",
+            );
+            const hasFormatProperty = Object.prototype.hasOwnProperty.call(json, "format");
+
+            if (!(hasPrettierProperty && hasPrettierFixProperty && hasFormatProperty)) {
+                return "Error parsing JSON file: " + file;
+            }
```
Comment (Josmithr) [inline build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:609]: We should explicitly return `undefined` at the end of this function for the success case.

**Review by Josmithr — CHANGES_REQUESTED** [review]
Looking good! My main feedback is about improving our error messaging, and adding a `resolver` to automatically "fix" the missing script issues.

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts [hunk 2]
```diff
@@ -585,6 +585,30 @@ export const handlers: Handler[] = [
...
+
+            const hasPrettierProperty = Object.prototype.hasOwnProperty.call(json, "prettier");
+            const hasPrettierFixProperty = Object.prototype.hasOwnProperty.call(
+                json,
+                "prettier:fix",
+            );
+            const hasFormatProperty = Object.prototype.hasOwnProperty.call(json, "format");
+
+            if (!(hasPrettierProperty && hasPrettierFixProperty && hasFormatProperty)) {
+                return "Error parsing JSON file: " + file;
```
[the remark this answers is not in the document]
Comment (jikim-msft) [reply build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts]: Return statement changed to return the `file` and list of missing scripts. Missing scripts are pushed to `missingScripts` array if it contains `false` value.

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts (hunk 1 above)
Comment (jikim-msft) [reply build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:624]: Added `return undefined` outside of the `if (!(hasPrettierProperty && hasPrettierFixProperty && hasFormatProperty))` clause.

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts [hunk 3]
```diff
@@ -585,6 +585,47 @@ export const handlers: Handler[] = [
...
+                json,
+                "prettier:fix",
+            );
+            const hasFormatProperty = Object.prototype.hasOwnProperty.call(json, "format");
+
+            if (!(hasPrettierProperty && hasPrettierFixProperty && hasFormatProperty)) {
+                //return "Error parsing JSON file: " + file;
+                const missingScripts: string[] = [];
+
+                if (hasPrettierProperty === false) {
```
Comment (Josmithr) [inline build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:611]: Since `hasPrettierProperty` is a boolean, this can be cleaned up as:
```suggestion
                if (!hasPrettierProperty) {
```
Unless I'm missing something.

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts [hunk 4]
```diff
@@ -585,6 +585,47 @@ export const handlers: Handler[] = [
...
+
+                if (hasPrettierFixProperty === false) {
+                    missingScripts.push(`prettier:fix`)
+                }
+
+                if (hasFormatProperty === false) {
+                    missingScripts.push(`format`)
+                }
+
+                return file + ` missing following scripts in package.json` + missingScripts.join("\n");
```
Comment (Josmithr) [inline build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:622]: Nit:
```suggestion
                return `${file} is missing the following scripts: ${missingScripts.join("\n\t")}`;
```

We generally recommend using the string interpolation syntax where possible. This file isn't doing a great job at that, but we might as well us it with new code.

I'm also suggesting adding a tab before each newline entry to make the formatting a bit nicer, but feel free to ignore / do what you think looks best :)

Comment (jikim-msft) [reply build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts]: Ah, I missed using string interpolation. Thank you for letting me know 👍

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts [hunk 5]
```diff
@@ -585,6 +585,46 @@ export const handlers: Handler[] = [
...
+            }
+
+            const hasPrettierProperty = Object.prototype.hasOwnProperty.call(json, "prettier");
+            const hasPrettierFixProperty = Object.prototype.hasOwnProperty.call(
+                json,
+                "prettier:fix",
+            );
+            const hasFormatProperty = Object.prototype.hasOwnProperty.call(json, "format");
+
+            if (!(hasPrettierProperty && hasPrettierFixProperty && hasFormatProperty)) {
```
Comment (tylerbutler) [inline build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:607]: I think you can remove this if-check since you're now checking them individually.

Comment (jikim-msft) [reply build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:607]: Then would I have to return error if the length of `missingScripts` is greater or equal to 1?

Comment (tylerbutler) [reply build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:607]: >0, I think.

Comment (tylerbutler) [reply build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:607]: Oh, that's the same, I guess. Ignore me. 🤦

**Review by Josmithr — CHANGES_REQUESTED** [review]

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts [hunk 6]
```diff
@@ -585,8 +585,107 @@ export const handlers: Handler[] = [
...
+        handler: (file) => {
+            let json;
+
+            try {
+                json = JSON.parse(readFile(file));
+            } catch (err) {
+                return "Error parsing JSON file: " + file;
+            }
+
+            const hasPrettierProperty = Object.prototype.hasOwnProperty.call(json, "prettier");
```
[the remark this answers is not in the document]
Comment (jikim-msft) [reply build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:600]: Change applied.

[the remark this answers is not in the document]
Comment (jikim-msft) [reply build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:600]: I think we should be calling the checks on `json.scripts`, since the `handler` is checking whether the specific scripts exist or not.

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts [hunk 7]
```diff
@@ -585,8 +585,107 @@ export const handlers: Handler[] = [
...
+    json: Map<any, any>,
+    jsonCheckRoot: string,
+    jsonDevDependencies: Map<string, string>,
+    jsonScripts: Map<string, string>,
+) {
+    // if devDependencies or scripts does not exist in package.json
+    if (jsonDevDependencies === undefined || jsonScripts === undefined) {
+        // create deveDependencies and add prettier
+        if (jsonDevDependencies === undefined) {
+            json["devDependencies"] = { prettier: "~2.6.2" };
```
[the remark this answers is not in the document]
Comment (jikim-msft) [reply build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:660]: Changed to `"prettier": undefined`

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts [hunk 8]
```diff
@@ -585,8 +585,107 @@ export const handlers: Handler[] = [
...
+        }
+
+        // create scripts and add format, prettier and prettier:fix scripts
+        if (jsonScripts === undefined) {
+            json["scripts"] = {
+                "format": "npm run prettier:fix",
+                "prettier": "prettier --check . --ignore-path ../../../.prettierignore",
```
[the remark this answers is not in the document]
Comment (jikim-msft) [reply build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts]: Change applied to discard after `--check .`

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (jikim-msft) [reply build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts]: Changed to `"prettier" : ""`

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (Josmithr) [reply build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts]: We should probably update all of the script checks to look under `json.scripts` then too, even though they are less likely to have conflicts under "dependencies", etc.

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts [hunk 9]
```diff
@@ -585,8 +585,107 @@ export const handlers: Handler[] = [
...
+    else {
+        jsonDevDependencies["prettier"] = "~2.6.2";
+
+        jsonScripts["format"] = "npm run prettier:fix";
+        jsonScripts["prettier"] = "prettier --check . --ignore-path ../../../.prettierignore";
+        jsonScripts["prettier:fix"] = "prettier --write . --ignore-path ../../../.prettierignore";
+    }
+
+    // package.json in root of the release-group should have lerna script
+    if (jsonCheckRoot === "root") {
```
[the remark this answers is not in the document]
Comment (tylerbutler) [reply build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:675]: No, this isn't a good way. That said, I don't know of a better one outside of the release tools context, which is way too heavy to load in a policy handler. Need to think about it some more...

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts [hunk 10]
```diff
@@ -585,8 +586,102 @@ export const handlers: Handler[] = [
...
+    jsonScripts["prettier"] = "prettier --check .";
+
+    jsonScripts["prettier:fix"] = "prettier --write .";
+
+    // package.json in root of the release-group should have lerna script
+    if (jsonCheckRoot === "root") {
+        jsonScripts["format"] = "lerna run format --no-sort --stream -- -- -- --color";
+    }
+
+    return undefined;
```
Comment (alexvy86) [inline build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:682]: Seems like this function doesn't need to return.

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts [hunk 11]
```diff
@@ -585,8 +586,102 @@ export const handlers: Handler[] = [
...
+
+            const hasPrettierScript = Object.prototype.hasOwnProperty.call(
+                json.scripts,
+                "prettier",
+            );
+            const hasPrettierFixScript = Object.prototype.hasOwnProperty.call(
+                json.scripts,
+                "prettier:fix",
+            );
+            const hasFormatScript = Object.prototype.hasOwnProperty.call(json.scripts, "format");
```
Comment (alexvy86) [inline build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:609]: I think the error you're seeing (`TypeError: Cannot convert undefined or null to object`) is coming from here, when analyzing a file that does not have a `scripts` section.

image

Do we need to do the check with that API for some reason, instead of `json.scripts?["prettier"]`?

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts [hunk 12]
```diff
@@ -585,8 +586,102 @@ export const handlers: Handler[] = [
...
+
+            if (!(hasPrettierScript && hasPrettierFixScript && hasFormatScript)) {
```
Comment (alexvy86) [inline build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:611]: I think this would be more readable without the outer `if`, and after the three individual checks something like `return missingScripts.length > 0 ? <the error> : undefined;`

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts [hunk 13]
```diff
@@ -585,8 +586,102 @@ export const handlers: Handler[] = [
...
+
+            return { resolved: resolved };
+        },
+    },
 ];
 
+function addPrettier(json: Record<string, any>) {
+    const jsonCheckRoot = json.name;
+    let jsonDevDependencies = json.devDependencies;
+    let jsonScripts = json.scripts;
```
Comment (alexvy86) [inline build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:653]: Do we need the two local variables? Feels like the rest of the function should be able to check/update `json["property-name"]` in place.

**Review by tylerbutler — COMMENTED** [review]
Some high-level feedback about the design and intent of this policy handler.

TL;DR: I think you can simplify quite a bit and be _less_ smart, and in this case I think that would be better.

1. I recommend making the resolver conservative. Don't just blast the scripts everywhere. Instead, maybe something like, "If the package has one of the prettier, format, or prettier:fix scripts, then all scripts should be present." The reason for this suggestion is that our repo is quite heterogeneous, and policy-check doesn't have many per-project configuration options. Thus, if the resolver is heavy-handed, you're more likely to "accidentally" affect projects you might not want to. Even if you think it should apply to every project, I predict that you'll find a few that you don't.
2. A corollary to 1: if a package has no scripts section, I wouldn't force add it. Essentially I am arguing for an "opt-in" model for enforcement. There doesn't seem to be an opt-out approach.
3. I wouldn't worry about checking the release group root package.jsons. There are very few release groups, and there's no good way to differentiate at the policy level between root packages and non-root. If we need such a thing, we should add it. (We could use the presence of a `fluidBuild` node as an indicator but build-tools doesn't have one right now.)

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts (hunk 10 above)
Comment (jikim-msft) [reply build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:682]: Removed `return undefined;`

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts (hunk 12 above)
Comment (jikim-msft) [reply build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:611]: Removed the outermost if-condition and replaced with ternary operator.

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts [hunk 14]
```diff
@@ -585,8 +586,93 @@ export const handlers: Handler[] = [
...
+            writeFile(file, JSON.stringify(sortPackageJson(json), undefined, 2) + newline);
+
+            return { resolved: resolved };
+        },
+    },
 ];
 
+function addPrettier(json: Record<string, any>) {
+    // if the json file contains scripts field
+    console.log(1);
```
Comment (alexvy86) [inline build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:648]: [image
](https://media3.giphy.com/media/8abAbOrQ9rvLG/giphy.gif?cid=ecf05e47b8ezjn4qjpuei8gf3uq49gh8rlsvs2yu9yhk85jz&rid=giphy.gif&ct=g)

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts [hunk 15]
```diff
@@ -585,8 +586,93 @@ export const handlers: Handler[] = [
...
+
+    const hasScriptsField = Object.prototype.hasOwnProperty.call(json, "scripts");
+
+    if (hasScriptsField) {
```
Comment (alexvy86) [inline build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:652]: I think this check is worth extracting out to the `resolver` above, so the `addPrettier` function always does exactly what its name suggests, instead of sometimes not doing anything. That would also save us the `writeFile()` call on line 639 which only needs to happen if the file was actually changed.

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts [hunk 16]
```diff
@@ -585,8 +586,93 @@ export const handlers: Handler[] = [
...
+        const hasPrettierScriptResolver = Object.prototype.hasOwnProperty.call(
```
Comment (alexvy86) [inline build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:653]: Is there an advantage to this way of doing the call over `json.scripts.hasOwnProperty("prettier")`? Or something like suggested in [this comment](https://github.com/microsoft/FluidFramework/pull/13072/files#r1036546558).

Comment (alexvy86) [reply build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts:653]: I see. I think I like `json.scripts?.prettier !== undefined` better anyway 😄 .

And aren't you supposed to be OOF @tylerbutler ? Shoo! 😛

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts (hunk 11 above)
Comment (jikim-msft) [reply build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts]: As @tylerbutler mentioned below, using it for the purpose of lint rule.

File: build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts (hunk 13 above)
Comment (jikim-msft) [reply build-tools/packages/build-tools/src/repoPolicyCheck/handlers/npmPackages.ts]: Both local variables removed.

**Comment by jikim-msft** [comment]
Made changes according to @alexvy86 @tylerbutler 's latest feedback. 

1. `Handler`
- Added `hasScriptsField` to check if each `package.json` contains `scripts` field. Initially, this check was not included and created `TypeError: Cannot convert undefined or null to object`
- If `missingScripts` has length greater or equal to 1, it will raise a warning, otherwise return `undefined`. 

2.`Resolver`
- Moved `hasScriptsField` from the helper function `addPrettier`.  With the modified clause, `addPrettier` and `writeFile` will be invoked if and only if the `package.json` contains `scripts` field. 

3. `AddPrettier`
- As @tylerbutler 's suggestion, `addPrettier` function takes a more conservative approach in terms of adding the scripts. First, it checks whether the `package.json` contains `format`, `prettier`, and `prettier:fix` script. If it has at least one script inside it, it will decide to include all three scripts. If not, this step will be ignored. 


I believe logic in `handler` and `resolver` are somewhat repetitive (since they are both initializing same boolean variable `hasScriptsField`). However, due to the way `handler` and `resolver` are separately structured, I believe `scripts` field check in both parts are inevitable. If there seems to be a better way, I would love any suggestions. 

In order to test this locally, run the following commands in:
1. `build-tools/packages/build-tools`: `npm run build`
2. `root`: `build-tools/packages/build-cli/bin/dev check policy --exclusions build-tools/packages/build-tools/data/exclusions.json --fix -D assert-short-codes`

**Review by alexvy86 — APPROVED** [review]
Seems ok to me that handler and resolver do their check for scripts separately.

**Comment by jikim-msft** [comment]
> Seems ok to me that handler and resolver do their check for scripts separately.

Great! Thanks for reviewing the PR!

**Review by Josmithr — APPROVED** [review]
Looks good to me!
